### PR TITLE
Introduce new parameter max_bytes to limit the size of the buffer.

### PR DESCRIPTION
**Pull request by thomasschickinger**
Move the logic for limiting the buffer size by max_lines and max_bytes
to the TraceAccumulator.

### CODE REVIEW

File: lib/fluent/plugin/exception_detector.rb [hunk 1]
```diff
@@ -190,34 +190,36 @@ class TraceAccumulator
     # message_field may contain the empty string. In this case, the
     # TraceAccumulator 'learns' the field name from the first record by checking
     # for some pre-defined common field names of text logs.
-    def initialize(message_field, languages, &emit_callback)
+    # The optional parameters max_lines and max_bytes limit the maximum amount
```
Comment (igorpeshansky) [inline lib/fluent/plugin/exception_detector.rb]: Nit: these aren't optional, they're named. To make them optional, you'd use the `name=value` syntax, i.e.:
```
def initialize(message_field, languages, max_lines=0, max_bytes=0, &emit_callback)
```
which would require you to pass `max_lines` and `max_bytes` as regular (unnamed) arguments, i.e.:
```
Fluent::TraceAccumulator.new(message, @languages, max_lines, max_bytes) do |t, r|
```
Since you intend to specify one or the other, I'd just s/optional/named/. :-)

File: lib/fluent/plugin/exception_detector.rb [hunk 2]
```diff
@@ -190,34 +190,36 @@ class TraceAccumulator
...
+    # of data to be buffered. The default value '0' indicates 'no limit'.
```
Comment (igorpeshansky) [inline lib/fluent/plugin/exception_detector.rb]: Nit: it's `0` (integer), not `'0'` (string).

File: test/plugin/test_exception_detector.rb [hunk 3]
```diff
@@ -350,4 +350,28 @@ def test_json_messages
...
+                                          [:all],
+                                          max_lines: max_lines) do |_, m|
+      out << m
+    end
+    feed_lines(buffer, JAVA_EXC)
+    assert_equal([JAVA_EXC_PART1] + JAVA_EXC_PART2.lines, out)
+  end
+
+  def test_max_bytes_limit
+    max_bytes = JAVA_EXC_PART1.length + JAVA_EXC_PART2.lines[0].length - 1
```
Comment (igorpeshansky) [inline test/plugin/test_exception_detector.rb]: You might want to add a comment, something along the lines of "_Make sure the trace is flushed after the first part._" (also helps in `test_max_lines_limit`, but here it's actually needed, IMO).
I would also test with `JAVA_EXC_PART1.length + 1`.

File: test/plugin/test_out_detect_exceptions.rb [hunk 4]
```diff
@@ -167,7 +167,7 @@ def test_flush_after_max_lines
       feed_lines(d, t, PYTHON_EXC, JAVA_EXC)
     end
     expected =
-      [PYTHON_EXC.lines[0..2].join] + PYTHON_EXC.lines[3..-1] + [JAVA_EXC]
+      [PYTHON_EXC.lines[0..1].join] + PYTHON_EXC.lines[2..-1] + [JAVA_EXC]
```
Comment (igorpeshansky) [inline test/plugin/test_out_detect_exceptions.rb]: Why was this change needed to the test? Is this because of the new defaults?

File: lib/fluent/plugin/exception_detector.rb (hunk 1 above)
Comment (thomasschickinger) [inline lib/fluent/plugin/exception_detector.rb]: Thanks, done.

File: lib/fluent/plugin/exception_detector.rb (hunk 2 above)
Comment (thomasschickinger) [inline lib/fluent/plugin/exception_detector.rb]: Done.

File: test/plugin/test_out_detect_exceptions.rb (hunk 4 above)
Comment (thomasschickinger) [inline test/plugin/test_out_detect_exceptions.rb]: The previous implementation counted the size of the buffer. So, when the limit was 2, the buffer could contain two lines and only the third line would trigger the flush, resulting in a log entry containing three lines. I figured that this might be surprising for users who don't care about our internal buffer but about the size of the log entries that they get.

I've adapted the description of the command-line parameter to make this clearer.

File: lib/fluent/plugin/out_detect_exceptions.rb [hunk 5]
```diff
@@ -30,8 +30,10 @@ class DetectExceptionsOutput < Output
     config_param :multiline_flush_interval, :time, default: nil
     desc 'Programming languages for which to detect exceptions. Default: all.'
     config_param :languages, :array, value_type: :string, default: []
-    desc 'Maximum number of lines in a detected stack trace. Default: 1000.'
+    desc 'Max. number of lines in combined log entries. Default: 1000.'
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_detect_exceptions.rb]: [Optional] This could also be "Maximum number of lines to flush", which would fit on one line... Also below.

File: test/plugin/test_exception_detector.rb [hunk 6]
```diff
@@ -363,7 +363,9 @@ def test_max_lines_limit
     assert_equal([JAVA_EXC_PART1] + JAVA_EXC_PART2.lines, out)
   end
 
-  def test_max_bytes_limit
+  def test_high_max_bytes_limit
+    # Limit is just too small to add one more line to the buffered first part of
+    # the exception.
     max_bytes = JAVA_EXC_PART1.length + JAVA_EXC_PART2.lines[0].length - 1
     out = []
     buffer = Fluent::TraceAccumulator.new(nil,
```
Comment (igorpeshansky) [inline test/plugin/test_exception_detector.rb]: Are there stray tabs in lines 372-373? Or are they just mis-indented?

File: test/plugin/test_exception_detector.rb (hunk 3 above)
Comment (igorpeshansky) [inline test/plugin/test_exception_detector.rb]: Let's also add this comment to `test_max_lines_limit`, for consistency.

File: test/plugin/test_exception_detector.rb [hunk 7]
```diff
@@ -350,4 +350,46 @@ def test_json_messages
       assert_equal(expected_json(s.actual_field, s.output), out, s.desc)
     end
   end
+
+  def test_max_lines_limit
+    max_lines = JAVA_EXC_PART1.lines.length
+    out = []
+    buffer = Fluent::TraceAccumulator.new(nil,
+                                          [:all],
+                                          max_lines: max_lines) do |_, m|
```
Comment (igorpeshansky) [inline test/plugin/test_exception_detector.rb]: Just realized: instead of relying on the default limit, you should specify "unlimited" for `max_bytes` here and `max_lines` below (especially true since you've now changed the default limit for `max_lines` from "unlimited").

File: test/plugin/test_out_detect_exceptions.rb (hunk 4 above)
Comment (igorpeshansky) [inline test/plugin/test_out_detect_exceptions.rb]: Thanks for adjusting the parameter descriptions... You might still add a comment here saying something like _Check that only the first 2 lines of the trace are kept_. It's also not entirely clear why you didn't also expect to detect the first 2 lines of the Java trace, which might merit another comment.

Comment (thomasschickinger) [inline test/plugin/test_out_detect_exceptions.rb]: No problem. Done.

File: lib/fluent/plugin/out_detect_exceptions.rb (hunk 5 above)
Comment (thomasschickinger) [inline lib/fluent/plugin/out_detect_exceptions.rb]: Done

File: test/plugin/test_out_detect_exceptions.rb [hunk 8]
```diff
@@ -166,6 +166,11 @@ def test_flush_after_max_lines
     d.run do
       feed_lines(d, t, PYTHON_EXC, JAVA_EXC)
     end
+    # Expected: the firsti two lines of the exception are buffered and combined.
```
Comment (igorpeshansky) [inline test/plugin/test_out_detect_exceptions.rb]: Typo: firsti.

File: test/plugin/test_out_detect_exceptions.rb [hunk 9]
```diff
@@ -166,6 +166,11 @@ def test_flush_after_max_lines
...
+    # Then the max_lines setting kicks in and the rest Python exception is
```
Comment (igorpeshansky) [inline test/plugin/test_out_detect_exceptions.rb]: the rest *of the* Python exception?

File: test/plugin/test_out_detect_exceptions.rb [hunk 10]
```diff
@@ -166,6 +166,11 @@ def test_flush_after_max_lines
...
+    # logged line-by-line (since it's not an exception stack in itself).
+    # Finally, the Java exception is logged in a single log entry, since it
```
Comment (igorpeshansky) [inline test/plugin/test_out_detect_exceptions.rb]: Ohh, that explains it.
[Optional] I would propose changing "in a single log entry" to "in its entirety" to make this even clearer.

File: test/plugin/test_exception_detector.rb (hunk 7 above)
Comment (igorpeshansky) [inline test/plugin/test_exception_detector.rb]: Sorry, I confused the various defaults.

File: test/plugin/test_exception_detector.rb (hunk 3 above)
Comment (thomasschickinger) [inline test/plugin/test_exception_detector.rb]: Done

File: test/plugin/test_out_detect_exceptions.rb (hunk 8 above)
Comment (thomasschickinger) [inline test/plugin/test_out_detect_exceptions.rb]: Done

File: test/plugin/test_out_detect_exceptions.rb (hunk 9 above)
Comment (thomasschickinger) [inline test/plugin/test_out_detect_exceptions.rb]: Done

**Comment by igorpeshansky** [comment]
LGTM :shipit:
